lib/msal-browser/src/cache/BrowserCacheManager.ts

```
@@ -286,6 +331,135 @@ export class BrowserCacheManager extends CacheManager {
+        };
+    }
+
+    /**
+     * Adds the given key to the token key map
+     * @param key 
+     * @param type 
+     */
+    addTokenKey(key: string, type: CredentialType): void {
+        this.logger.trace("BrowserCacheManager addTokenKey called");
+        const tokenKeys = this.getTokenKeys();
+
+        switch (type) {
+            case CredentialType.ID_TOKEN:
+                if (tokenKeys.idToken.indexOf(key) === -1) {
+                    this.logger.info("BrowserCacheManager: addTokenKey - idToken added to map");
+                    tokenKeys.idToken.push(key);
+                }
+                break;
+            case CredentialType.ACCESS_TOKEN:
+                if (tokenKeys.accessToken.indexOf(key) === -1) {
+                    this.logger.info("BrowserCacheManager: addTokenKey - accessToken added to map");
+                    tokenKeys.accessToken.push(key);
+                }
+                break;
+            case CredentialType.REFRESH_TOKEN:
+                if (tokenKeys.refreshToken.indexOf(key) === -1) {
+                    this.logger.info("BrowserCacheManager: addTokenKey - refreshToken added to map");
+                    tokenKeys.refreshToken.push(key);
+                }
+                break;
+            default:
+                this.logger.error("BrowserCacheManager:addTokenKey - CredentialType provided invalid. Doing nothing.");
+        }
+
+        this.setItem(`${StaticCacheKeys.TOKEN_KEYS}.${this.clientId}`, JSON.stringify(tokenKeys));
+    }
+
+    /**
+     * Removes the given key from the token key map
+     * @param key 
+     * @param type 
+     */
+    removeTokenKey(key: string, type: CredentialType): void {
+        this.logger.trace("BrowserCacheManager removeTokenKey called");
+        const tokenKeys = this.getTokenKeys();
+
+        switch (type) {
+            case CredentialType.ID_TOKEN:
+                const idRemoval = tokenKeys.idToken.indexOf(key);
+                if (idRemoval > -1) {
```

Are we swallowing the errors if the given key is not present in the map? Should we at least log it? Same applies below!

---

Logging added